src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAdminController.java

```diff
@@ -142,10 +151,63 @@ public void fullDeleteStudy(Long[] ids) {
             UserSearch userSearch)
             throws UserNotFoundException {
 
-        Page<IUser> users = userService.getAll(pageable, userSearch);
+        Page<U> users = userService.getAll(pageable, userSearch);
         return users.map(user -> conversionService.convert(user, CommonUserDTO.class));
     }
 
+    @ApiOperation("Register new users")
+    @RequestMapping(value = "/api/v1/admin/users/group", method = RequestMethod.POST)
+    public void register(@RequestBody BulkUsersRegistrationDTO bulkUsersDto) throws PasswordValidationException, ValidationException {
+
+        if (bulkUsersDto.getTenantIds() == null || bulkUsersDto.getTenantIds().isEmpty()) {
+            throw new ValidationException("tenants: must be not empty");
+        }
+
+        boolean emailConfirmationRequired = bulkUsersDto.getEmailConfirmationRequired();
+
+        Set<Tenant> tenants = bulkUsersDto.getTenantIds().stream()
+                .map(tenantId -> conversionService.convert(tenantId, Tenant.class))
+                .collect(Collectors.toSet());
+        List<U> users = bulkUsersDto.getUsers().stream()
+                .map(dto -> (U) conversionService.convert(dto, User.class))
+                .collect(Collectors.toList());
+        updateFields(users, tenants, emailConfirmationRequired, bulkUsersDto.getPassword());
+
+        List<U> createdUsers = userService.createAll(users);
+
+        if (emailConfirmationRequired) {
+            for (U user : createdUsers) {
+                bulkUsersDto.getUsers().stream().forEach(userDto -> {
+                    if (userDto.getEmail().equals(user.getEmail())) {
```

Comment:
Can be moved to .filter() of the stream

Reply:
Moreover probably it`s better to convert UserDTO list to map with email as key instead of filtering collection for each user